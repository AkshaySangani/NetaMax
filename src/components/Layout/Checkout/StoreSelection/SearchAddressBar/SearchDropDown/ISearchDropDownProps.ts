import { Status } from "use-places-autocomplete";

export interface ISearchDropDownProps {
  /**
   * List of recommended addresses
   * @type {google.maps.places.AutocompletePrediction[]}
   */
  data: google.maps.places.AutocompletePrediction[];

  /**
   * Status
   * @type {Status}
   */
  status: Status;

  /**
   * Handles the select of a searched address
   * @type {(val: string, showVal: string) => void}
   */
  handleSelectAddress: (val: string, showVal: string) => void;
}
