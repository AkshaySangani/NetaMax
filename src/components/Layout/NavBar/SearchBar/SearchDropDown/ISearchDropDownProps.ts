import { IProduct } from "dataflows/Product/IProduct";

export interface ISearchDropDownProps {
  /**
   * The input value.
   * @type {string}
   */
  value?: string;

  /**
   * The product list.
   * @type {IProduct[]}
   */
  products?: IProduct[];

  /**
   * Indicates if the search products are loading
   * @type {boolean}
   **/
  isLoadingSearchProduct: boolean;

  /**
   * The empty search message
   * @type {string}
   */
  emptySearchMessage: string;

  /**
   * Indicates if the search has more products
   * @type {boolean}
   **/
  hasMoreProducts: boolean;

  /**
   * Clear the search products
   * @type {() => void}
   **/
  clearSearchProducts: () => void;

  /**
   * The add to cart event handler.
   * @type {(product: IProduct) => void}
   */
  onAddToBasketClick: (product: IProduct) => void;

  /**
   * The remove from cart event handler.
   * @type {(product: IProduct) => void}
   */
  onRemoveFromBasketClick: (product: IProduct) => void;

  /**
   * Gets the quantity of the product.
   * @type {(product: IProduct) => number}
   */
  getQtyInBasket: (product: IProduct) => number;

  /**
   * Handles "Ver mÃ¡s" click
   * @type {() => void}
   **/
  handleSearchProducts: () => void;
}
