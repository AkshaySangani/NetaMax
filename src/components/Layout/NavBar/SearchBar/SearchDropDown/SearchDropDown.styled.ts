import styled from "styled-components";
import { responsive } from "styled/responsive";

import { Box } from "@chakra-ui/react";

export const StyledDropdown = styled(Box)`
  position: absolute;
  top: ${(props) => (props.$isLargerThanSm ? "0" : "0")};
  left: ${(props) => (props.$isLargerThanSm ? "none" : "0")};
  width: ${(props) => (props.$isLargerThanSm ? "480px" : "100%")};
  height: 50px;
  background-color: #fff;
  margin-top: 55px;
  box-shadow: rgba(0, 0, 0, 0.1) 0 0 0 0, rgba(0, 0, 0, 0.1) 0px 4px 11px;
  border-radius: 5px;
  z-index: 10;
  > ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
    > li {
      padding-right: 10px;
      padding-left: 2px;
      justify-content: center;
      align-items: center;
      background-color: #fff;
      border-bottom: 0.5px solid #cbcbcb;
      height: 60px;
      display: flex;
      align-items: center;
      a {
        text-decoration: none;
        -webkit-box-shadow: none;
        -moz-box-shadow: none;
        box-shadow: none;
      }
      &:hover {
        cursor: pointer;
      }
      &:last-child {
        border-bottom-left-radius: 12px;
        border-bottom-right-radius: 12px;
        border-bottom: 0px;
      }
      &:first-child {
        ${responsive.smPlus} {
          border-top-left-radius: 12px;
          border-top-right-radius: 12px;
        }
      }
      &:not(:first-child) {
        border-top: 0;
      }
    }
  }
`;
