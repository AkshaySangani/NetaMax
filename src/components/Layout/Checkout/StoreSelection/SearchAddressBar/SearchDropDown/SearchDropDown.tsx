import { ReactElement } from "react";

import { Loader } from "components/common/Loader";
import { AddressCirclePin } from "styled/icons/CheckoutFlow/findStoreWithMap/AddressCirlcePin";
import { ChevronRight } from "styled/icons/CheckoutFlow/findStoreWithMap/ChevronRight";

import { Box, Center, Flex, Text, VStack } from "@chakra-ui/react";

import { ISearchDropDownProps } from "./ISearchDropDownProps";
import { PLACES_STATUS_OK, PLACES_STATUS_ZERO_RESULTS } from "constants/checkoutConstants";

/**
 * The SearchDropDown component.
 * @param { ISearchDropDownProps } props the component props.
 * @returns { ReactElement } The SearchDropDown component.
 */
export const SearchDropDown = (props: ISearchDropDownProps): ReactElement => {
  const { data, status, handleSelectAddress } = props;

  return (
    <VStack width="100%" zIndex="10" backgroundColor="white" paddingX="16px">
      {status === PLACES_STATUS_OK ? (
        data.map((address, i) => (
          <Flex
            onClick={() =>
              handleSelectAddress(address.description, address.structured_formatting.main_text)
            }
            marginTop="0px"
            borderBottom="1px solid rgba(0, 0, 0, .03)"
            paddingY="16px"
            width="100%"
            key={`${address.description}-${i}`}
            justifyContent="space-between"
            alignItems="center"
            cursor="pointer"
          >
            <Flex alignItems="center">
              <AddressCirclePin />
              <Flex marginLeft="8px" direction="column">
                <Flex>
                  <Text fontWeight="600" isTruncated maxWidth="280px">
                    {address.structured_formatting.main_text}
                  </Text>
                </Flex>
                <Text opacity="0.7" isTruncated maxWidth="280px">
                  {address.structured_formatting.secondary_text}
                </Text>
              </Flex>
            </Flex>
            <ChevronRight />
          </Flex>
        ))
      ) : status === PLACES_STATUS_ZERO_RESULTS ? (
        <Box textAlign="center" paddingY="16px">
          <Text fontWeight="600">No se encontraron resultados para la direcci√≥n ingresada</Text>
        </Box>
      ) : (
        <Center height="150px">
          <Loader size="md" delay={300} />
        </Center>
      )}
    </VStack>
  );
};
