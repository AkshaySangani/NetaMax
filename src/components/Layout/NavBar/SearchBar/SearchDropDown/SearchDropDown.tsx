import { ReactElement } from "react";

import { AiOutlineArrowRight } from "react-icons/ai";

import { IncrementalButton } from "components/common/IncrementalButton/IncrementalButton";
import { Loader } from "components/common/Loader";
import { CURRENCY_NAME } from "constants/localeFormatConstants";
import NextLink from "next/link";
import { screenSizes } from "styled/screen";
import { formatMoney } from "utils/currencyUtils";
import { productIsNotAvailable, productMaxQuantity } from "utils/productUtils";

import {
  useMediaQuery,
  Box,
  Button,
  Center,
  Flex,
  Image,
  Link,
  Spacer,
  Text,
} from "@chakra-ui/react";

import { ISearchDropDownProps } from "./ISearchDropDownProps";
import { StyledDropdown } from "./SearchDropDown.styled";

/**
 * The SearchDropDown Component.
 * @param {ISearchDropDownProps} props the component props.
 * @returns {ReactElement} the searchDropDown component.
 */
export const SearchDropDown = (props: ISearchDropDownProps): ReactElement => {
  const {
    value,
    products,
    isLoadingSearchProduct,
    onAddToBasketClick,
    getQtyInBasket,
    onRemoveFromBasketClick,
    clearSearchProducts,
    handleSearchProducts,
    hasMoreProducts,
    emptySearchMessage,
  } = props;

  const [isLargerThanSm] = useMediaQuery(`(min-width: ${screenSizes.smMin}px)`);

  if (!value || (value && value?.length <= 2)) {
    return <></>;
  }
  return isLoadingSearchProduct ? (
    <StyledDropdown $isLargerThanSm={isLargerThanSm}>
      <Center h="100%" w="100%">
        <Loader size="sm" /> Buscando...
      </Center>
    </StyledDropdown>
  ) : (
    <StyledDropdown $isLargerThanSm={isLargerThanSm}>
      {emptySearchMessage !== "" ? (
        <Center h="100%" w="100%">
          <Text>{emptySearchMessage} ðŸ˜”</Text>
        </Center>
      ) : (
        <ul>
          {products &&
            products.map((product) => {
              const qty = getQtyInBasket(product);
              return (
                <li key={product.id}>
                  <Flex width="60%">
                    <NextLink href={`/products/${product.id}`} passHref>
                      <Link
                        onClick={clearSearchProducts}
                        _focus={{
                          boxShadow: "0 0 1px 2px transparent",
                        }}
                        textDecor="none"
                      >
                        <Flex
                          alignItems="center"
                          maxWidth={{ base: qty > 0 ? "200px" : "224px", md: "100%" }}
                        >
                          <Image
                            src={product.seoFilename}
                            height="45px"
                            w="45px"
                            minW={"45px"}
                            fallbackSrc="/assets/images/noimg.png"
                            borderRadius="8px"
                            boxShadow="0px 0px 9px rgb(18 17 17 / 15%)"
                            marginLeft="6px"
                            marginRight="10px"
                            padding="1px"
                          />
                          <Flex flexDir="column">
                            <Text
                              isTruncated
                              fontSize="sm"
                              color="#6A696E"
                              maxWidth={{
                                base: qty > 0 ? "180px" : "224px",
                                md: "300px",
                              }}
                            >
                              {product.name}
                            </Text>
                            <Flex>
                              <Text fontWeight={700}>
                                {formatMoney(parseFloat(product.price), CURRENCY_NAME)}{" "}
                              </Text>
                              {product.oldPrice > product.price && (
                                <Text pl={2} color="gray.500" fontSize="xs" as="del">
                                  {formatMoney(parseFloat(product.oldPrice), CURRENCY_NAME)}
                                </Text>
                              )}
                            </Flex>
                          </Flex>
                        </Flex>
                      </Link>
                    </NextLink>
                    <Box></Box>
                  </Flex>
                  <Spacer />
                  <Flex width={qty > 0 ? "35%" : "15%"} justifyContent="flex-end">
                    {qty > 0 ? (
                      <IncrementalButton
                        min={product.orderMinimumQuantity}
                        max={productMaxQuantity(
                          product.manageInventoryMethodId,
                          product.stockQuantity,
                          product.orderMaximumQuantity
                        )}
                        defaultValue={qty}
                        onAdd={() => onAddToBasketClick(product)}
                        onSubtract={() => onRemoveFromBasketClick(product)}
                      />
                    ) : (
                      <Button
                        data-testid={`add-button-${product.id}`}
                        disabled={productIsNotAvailable(
                          product.manageInventoryMethodId,
                          product.stockQuantity
                        )}
                        onClick={() => {
                          onAddToBasketClick(product);
                        }}
                        height={{ base: "30px", md: "30px" }}
                        width={{ base: "60px", md: "60px" }}
                        _hover={{ bgColor: "none" }}
                        _disabled={{ bgColor: "netaGray.300", cursor: "no-drop" }}
                        colorScheme="netaBlue"
                        borderRadius="full"
                        fontSize="0.8rem"
                      >
                        AÃ±adir
                      </Button>
                    )}
                  </Flex>
                </li>
              );
            })}
          {hasMoreProducts && (
            <li style={{ backgroundColor: "#f3f3f3" }}>
              <Button
                data-testid="hasMoreProductsButton"
                borderRadius="full"
                color="netaBlue.500"
                bg="#E2E2E2"
                onClick={() => {
                  handleSearchProducts();
                }}
              >
                Ir a ver mÃ¡s resultados <AiOutlineArrowRight />
              </Button>
            </li>
          )}
        </ul>
      )}
    </StyledDropdown>
  );
};
