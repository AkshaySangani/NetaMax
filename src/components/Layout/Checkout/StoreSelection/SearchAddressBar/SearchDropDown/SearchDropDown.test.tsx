import { cleanup, render, screen } from "@testing-library/react";
import { ISearchDropDownProps } from "./ISearchDropDownProps";
import { SearchDropDown } from "./SearchDropDown";

describe("Render SearchDropDown component", () => {
  let dataMocked: google.maps.places.AutocompletePrediction[];
  let mock: ISearchDropDownProps;
  beforeEach(() => {
    dataMocked = [
      {
        description: "Adana, Reşatbey, Seyhan/Provincia de Adana, Turquía",
        distance_meters: 12133677,
        matched_substrings: [],
        place_id: "ChIJk3GcpziPKBURBuoLWNh0FtY",
        reference: "ChIJk3GcpziPKBURBuoLWNh0FtY",
        structured_formatting: {
          main_text: "Adana",
          main_text_matched_substrings: Array(1),
          secondary_text: "Reşatbey, Seyhan/Provincia de Adana, Turquía",
        },
        terms: [],
        types: ["locality", "political", "geocode"],
      },
    ];
    mock = {
      data: dataMocked,
      status: "OK",
      handleSelectAddress: jest.fn(),
    };
  });

  afterEach(() => {
    cleanup();
  });

  test("SearchDropDown should be rendered correctly", () => {
    const { container } = render(<SearchDropDown {...mock} />);
    expect(container).toMatchSnapshot();
  });

  test("Main text of data should be rendered properly", () => {
    render(<SearchDropDown {...mock} />);
    const dataList = mock.data.length > 0 ? mock.data : [];
    const mainText = dataList[0]?.structured_formatting.main_text || "this text is not in the doc";
    expect(screen.getByText(mainText)).toBeInTheDocument();
  });

  test("Secondary text of data should be rendered properly", () => {
    render(<SearchDropDown {...mock} />);
    const dataList = mock.data.length > 0 ? mock.data : [];
    const secondaryText =
      dataList[0]?.structured_formatting.secondary_text || "this text is not in the doc";
    expect(screen.getByText(secondaryText)).toBeInTheDocument();
  });
});
