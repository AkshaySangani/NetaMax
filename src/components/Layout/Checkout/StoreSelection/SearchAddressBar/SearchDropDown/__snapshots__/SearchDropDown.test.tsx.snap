// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render SearchDropDown component SearchDropDown should be rendered correctly 1`] = `
<div>
  <div
    class="chakra-stack css-b71nz5"
  >
    <div
      class="css-1h369zo"
    >
      <div
        class="css-70qvj9"
      >
        <svg
          fill="none"
          height="32"
          viewBox="0 0 32 32"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="16"
            cy="16"
            fill="#353535"
            fill-opacity="0.04"
            r="16"
          />
          <path
            d="M16 15.3327C16.7363 15.3327 17.3333 14.7357 17.3333 13.9993C17.3333 13.263 16.7363 12.666 16 12.666C15.2636 12.666 14.6666 13.263 14.6666 13.9993C14.6666 14.7357 15.2636 15.3327 16 15.3327Z"
            stroke="#353535"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-opacity="0.4"
          />
          <path
            d="M16 22C16 22 20.6667 18.7692 20.6667 14.6154C20.6667 13.3913 20.175 12.2174 19.2999 11.3518C18.4247 10.4863 17.2377 10 16 10C14.7624 10 13.5754 10.4863 12.7002 11.3518C11.825 12.2174 11.3334 13.3913 11.3334 14.6154C11.3334 18.7692 16 22 16 22Z"
            stroke="#353535"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-opacity="0.4"
            stroke-width="1.5"
          />
        </svg>
        <div
          class="css-1qemygu"
        >
          <div
            class="css-k008qs"
          >
            <p
              class="chakra-text css-1nfcp0q"
            >
              Adana
            </p>
          </div>
          <p
            class="chakra-text css-9lrgos"
          >
            Reşatbey, Seyhan/Provincia de Adana, Turquía
          </p>
        </div>
      </div>
      <svg
        fill="none"
        height="10"
        viewBox="0 0 6 10"
        width="6"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.70711 0.292893C1.31658 -0.0976311 0.683417 -0.0976311 0.292893 0.292893C-0.0976311 0.683417 -0.0976311 1.31658 0.292893 1.70711L1.70711 0.292893ZM5 5L5.70711 5.70711C6.09763 5.31658 6.09763 4.68342 5.70711 4.29289L5 5ZM0.292893 8.29289C-0.0976311 8.68342 -0.0976311 9.31658 0.292893 9.70711C0.683417 10.0976 1.31658 10.0976 1.70711 9.70711L0.292893 8.29289ZM0.292893 1.70711L4.29289 5.70711L5.70711 4.29289L1.70711 0.292893L0.292893 1.70711ZM4.29289 4.29289L0.292893 8.29289L1.70711 9.70711L5.70711 5.70711L4.29289 4.29289Z"
          fill="#009DB5"
        />
      </svg>
    </div>
  </div>
</div>
`;
